Warp Using New FilterAlgo

danieldresser opened the pull request:
Not ready for merging yet, ( we'll probably merge it once I've got Warp actually using it ), but adding for comment before I go farther.  Still needs some cleanup, and I there's some spots that look like they need optimization but I suppose I should probably run a profile before doing anything about it.

One test case just checks a bunch of different derivative directions against an image I output - not based on anything, but it seems to show reasonable behaviour.

The other test case validates it against Resample, which works perfectly in all cases except sinc, which is where Resample seems to be wrong.  I'm not sure whether we want to bother changing this, but as I was thinking about how to present createFilter in FilterAlgo, I thought it might be worth showing just how broken the current approach can look.

danieldresser commented:
Hmm, was going to upload png versions of that failing test, but it seems that I can't attach images at the moment.  I'll show you tomorrow.

johnhaddon commented on src/GafferImage/FilterAlgo.cpp:
```diff
@@ -0,0 +1,222 @@
+			V2f filterPos = offset[0] * xstep + offset[1] * ystep;
+			float w = (*filter)( filterPos[0], filterPos[1] );
+
+			// \todo : I can't think of any way to keep this around cleanly for testing, since
```
Is this pretty much equivalent to a disk filter? Could you get the same effect by passing a disk filter in from the outside? Or is it important to be `0.1f` rather than `0.0f` outside the filter radius?

danieldresser replied:
Having a background value on the outside allows you to see where the edges of the bounding box are - if the bounding box is too big, the tests will all pass, but it's inefficient.  I'm not too concerned with removing this code once I've got everything working though - it's not an area where regressions are likely once I've got things set up properly.

johnhaddon commented on src/GafferImage/FilterAlgo.cpp:
```diff
@@ -0,0 +1,222 @@
+	{
+		V2f cp = p + (( corner & 1 ) ? 1.0f : -1.0f ) * hdpdx + (( corner & 2 ) ? 1.0f : -1.0f ) * hdpdy;
+
+		maxX = std::max( maxX, (int)floorf( cp[0] ) );
```
Can you explain why we're using floor when computing the max, and ceil when computing the min? Doesn't that mean we're ignoring pixels where the filter would be non-zero?

danieldresser replied:
I'm fairly confident it is working, but it is kinda tough to explain at the moment how this stuff lines up.  I forgot to put a todo in, but I think I need to rejigger where some 0.5 offsets to the pixel center are applied in order to make this stuff make more sense, and then I'll be in a better position to explain this.

danieldresser replied:
Actually, maybe it's easiest to explain with the pixel centers at integer values like is currently:

We evaluate the filter at pixel centers.  If the filter parallelogram does not cover a pixel center, we don't need to consider that pixel.  If the parallelogram is 0 sized, it includes 0 pixels.

So in order to actually get a pixel in column x, the filter parallelogram must extend above x ( so that floor( curCorner.x ) >= x, and also extend below x ( so that ceil( curCorner.x ) <= x ).

The bounding range is inclusive, so this calculation does actually include all pixels where the filter is non-zero at the center, even though the floors/ceils look kinda backward.

The test that I used for this is to turn on that commented out code that replaces the filter weight with a constant value for the whole bounding box, and another value inside the filter, and make sure that bounding box isn't larger than needed, and also a test that arbitrarily making the bounding size larger doesn't change test results.  Neither of these is particularly easy to turn into a unit test, since they would require exposing weird flags to hack with the internal behaviour.

Your suggestion to switch to Box.extend here is good, I'll do that.

Even though I'm finding it easier to think about this in term of pixel centers at integer coordinates, I should probably add a -0.5 to the position at the top of the function, since it appears we've got a convention of pixel centers at 0.5?

johnhaddon replied:
Ah, I see, I think. Are we overcooking the bound in Resample then? And yep, centres are at 0.5, edges are integers, so let's formulate everything that way so it's easier to follow. Then hopefully with an explanatory comment that section will be more evidently right.

johnhaddon commented on src/GafferImage/FilterAlgo.cpp:
```diff
@@ -0,0 +1,222 @@
+	const char *filterName = name.c_str();
+	if( name == "" )
+	{
+		filterName = "blackman-harris"; // TODO - if we don't know whether we're scaling up or down, should the default be more or less aggressive?
```
I'd say if we have no information, we shouldn't be guessing at all.

johnhaddon commented on src/GafferImage/FilterAlgo.cpp:
```diff
@@ -0,0 +1,222 @@
+// include another pixel, it switches on suddenly rather than fades up from
+// black. This variant approaches very close to zero at the edge of its
+// support, making it more suitable for use in a blur.
+class SmoothGaussian2D : public OIIO::Filter2D
```
I see this is copied over from Resample, but that you haven't removed it from there. Can we rejig things so we don't have duplicates? Maybe we can make `FilterAlgo::createFilter()` usable from Resample too somehow? Perhaps keeping just the logic for upsize/downsize filter defaults in Resample?

johnhaddon commented on src/GafferImage/FilterAlgo.cpp:
```diff
@@ -0,0 +1,222 @@
+		for( int x = minX; x <= maxX; x++ )
+		{
+
+			// \todo : I want to remove some pointless multiplies from here
```
Is the idea that the filter is already doing some multiplies, so we could just pre-adjust the filter radii instead? If so, I think we could benefit from the same thing in Resample. Maybe we can reformulate `createFilter()` to make this all a bit more straightforward?

Since I was a bit out of my depth and cribbing from OIIO when I wrote Resample, I didn't deviate too much from their example, but I did wonder if it couldn't be reformulated so all the multiplies were in the filter. Actually, I think maybe it can't be because of sinc? Because sinc is windowed? But since you seem suspicious of Resample/sinc anyway, perhaps you have other ideas?

danieldresser replied:
So, in terms of the behaviour of Resample/sinc, I'm not sure whether you've seen just how bad this can get currently - I can't seem to upload images at the moment, so I've just temporarily added a commit with  resamplePatterns_300x300_sinc.png and resamplePatterns_300x300_sinc.png.  The first is the current behaviour of this code, the second is the behaviour of Resample ( and OIIO ).  Hopefully these images make a convincing argument that if you're going to use a filter for resizing, you need to scale the filter, not just change the support width.  In OIIO's implementation of every filter but sinc, changing the width parameters does just scale it.

If we're happy with the idea of never changing the width of the OIIO filters, and only applying scaling on our end, then this definitely does open up some room to clean up some of this stuff.

johnhaddon replied:
> Hopefully these images make a convincing argument that if you're going to use a filter for resizing, you need to scale the filter, not just change the support width.

Indeed they do! Although I'm still a bit confused as to what is going wrong - is it just the same as [this](http://lists.openimageio.org/pipermail/oiio-dev-openimageio.org/2015-October/000168.html)?. In any case, do you think you could fix the sinc Resample (while keeping all other output the same) as part of this branch?

> If we're happy with the idea of never changing the width of the OIIO filters, and only applying scaling on our end, then this definitely does open up some room to clean up some of this stuff.

We currently do have the option of overriding filter width in the Resample node - it's needed to be used internally by the Blur node.

danieldresser replied:
> is it just the same as this?
Seems similar - basically OIIO API doesn't seem to have any definition of what "width" of a filter means - changing the width just gets you a filter that is different somehow.  In particular, there is no guarantee that a high width filter is blurrier than a low width filter ( it just happens to be true for all filters but sinc ).

I would argue that for both Resample and Blur, we should not be touching the width passed to OIIO.  We should always use the default one, and then apply our own scaling to the coordinates we sample the filter with.  This would have the desired effect of yielding identical results for all current filters, except for sinc, where it would match what I'm doing now with FilterAlgo.

johnhaddon replied:
Even for downsampling? I'm no expert, but my impression is sinc should be windowed for that?

johnhaddon commented on python/GafferImageTest/FilterAlgoTest.py:
```diff
@@ -0,0 +1,211 @@
+		sourceNet = redDotNetwork( dim )
+		s = GafferImage.Sampler( sourceNet[0]["out"], "R", sourceNet[0]["out"]["dataWindow"].getValue(), GafferImage.Sampler.BoundingMode.Black )
+	
+		filters = [ "", "box", "blackman-harris", "lanczos3", "radial-lanczos3", "sinc", "triangle", "gaussian", "smoothGaussian" ]
```
Seems like a good reason to have FilterAlgo::filters.

danieldresser commented on include/GafferImage/FilterAlgo.h:
```diff
@@ -0,0 +1,61 @@
+namespace FilterAlgo
+{
+
+typedef boost::shared_ptr<OIIO::Filter2D> Filter2DPtr;
```
Yeah, this sounds good to me.  The only question is whether we would ever want to change the default support of something like sinc.  If the goal is "we want to exactly support everything OIIO does", then maybe we need to, but if the goal is just "we want to offer some decent filters", then always using the default support width seems fine, and then we could use an opaque type.

Or for that matter, just use an integer?  If we have a fixed number of filters we support, then we could just pass an index into a global list.

danieldresser commented:
Made some comments here, it does seem like it would help to have a chat about the direction we want to take for unifying this stuff.  I'll guess I'll try and catch you briefly before or after the meeting tomorrow.

johnhaddon commented:
Made a couple of comments myself - let's see if we can fit in a chat after today's meeting then.

danieldresser commented:
Before I dashed out of work, I pushed up the commits I've got that are currently working - I think I'm now getting correct results from Warp.  I wouldn't necessarily bother with much review yet - before that, I should make sure I've addressed your previous comments, and squash a bunch of stuff to clean up history.  Names of various are also not particularly final.

Next on my list is to unify how Resample handles filters, and then double check the actual quality of the warps ( it looks good in simple cases, but I want to verify some production len distorts against Nuke ), and then try and clean up some terrible performance.  First step will be to profiled it, but I have some ideas where things may be going wrong.

The main thing I would appreciate you looking at is the current state of Warp.h / Warp.cpp.  Currently, I've added two extra layers of caching: positionPlug is an image containing sample positions, created by sampling the Engine at every point in the data window of the out plug, and positionsAndDerivativesPlug, which is the same thing, but including the 4 partial derivatives in x/y that are used for computing a filter parallelogram.  What is likely expensive: I've kept that outPlug depends on the hash of the sampler.  This means computing the input window of the transform for the tile.  That requires computing the value of positionsAndDerivativesPlug, which means computing the value of positionPlug in 5 tiles so that I have adjacency data for the derivatives, just to get a hash.  ( It probably doesn't help that I have to hit positionsAndDerivativesPlug 6 times for all the different channels ).  I wonder if it would be better to just use a pessimistic hash that says that all output tiles depend on the hash of all input tiles, which would save all this computation happening in the hash?

Things I'm thinking about trying:

a) Decide that in practice we probably won't have analytic warps that need to manually compute derivatives.  The numeric derivatives are probably fine.  This means that positionsAndDerivativesPlug is purely an internal cache, and doesn't need to make sense to anyone outside ( meaning I should probably put a "__" in the name and make the accessor protected.  The first thing that this means is that ensureMinimumParallelogramWidth can be done on the derivatives before they are stored in the cache, instead of considering the cache to be something that users of Gaffer could put potentially unsafe data in, and ensuring this in both hash and compute.
b) Convert positionsAndDerivativesPlug into not being an image plug.  It is something like an image plug, in that it returns a tile sized array of floats that depends on "image:tileOrigin" in the context, but if it's just an internal cache, then it doesn't need to be an image, which means it could return all data at once, rather than requiring 6 separate calls with different context settings for the channel.
c) If we don't decide to do a fully pessimistic hash where all input tiles affect all output tiles, then the input window bounds should also be stored on positionAndDerivativesPlug along with the per pixel data, avoiding a call to filteredTileBound that traverses all the pixels inside the hash ( which seems really bad )

I'm not sure what the type of positionAndDerivativesPlug should be if you agree it doesn't need to be an image?  It could be a CompoundObject, but unnecessary hash map indexing feels a bit wasteful?  If it's purely internal, I guess it could just be a FloatVectorData of length "6 * tileSize * tileSize + 4 " ( for 6 channels of data and the bounds ), but I can't imagine you'd like the obscurity of that.  It could be a custom struct type, but turning that into Cortex TypedData so I can stuff it in TypedPlug would be a bit verbose.  Maybe CompoundObject is the right answer.

The other overall question on Warp is whether the 3 separate cache layers ( enginePlug, positionPlug, positionAndDerivativePlug ) are all necessary.  Some thoughts - part of the reasoning you mentioned for Engine is to allow in the future using manipulators through a Warp.  This is only going to make sense for very well-behaved warps - maybe for a lens distort, or a rotate if we chose to do rotation this way, but not for some weird arbitrary UV warp that maps one source pixel to 12 different places in the output.  To me, in practice, there are two kinds of Warps:
* analytic warps : can be defined by a single engine for all tiles.  May support viewing upstream manipulators
* pixel defined warps: depend on input image data.  Vary per tile.  Probably not invertible, and probably not useful to support upstream manipulators.  Easiest to define by overriding compute() on positionsPlug

At this point, it feels like maybe the base Warp class is for pixel defined warps, and then maybe a small subclass for analytic warps that adds Engine to automatically drive positionsPlug, and can support your future manipulator trickery.  I'm just brainstorming here.

johnhaddon commented:
> Before I dashed out of work, I pushed up the commits I've got that are currently working - I think I'm now getting correct results from Warp.

Nice one!

> I wouldn't necessarily bother with much review yet - before that, I should make sure I've addressed your previous comments, and squash a bunch of stuff to clean up history. Names of various are also not particularly final.

Cool, I've skimmed through to get the gist, but refrained from making any comments. Actually, let me make just one now. I don't understand why the FilterAlgo namespace has extra stuff in it now - at first glance it feels like the internals of the Warp have just been exposed unnecessarily.

> The main thing I would appreciate you looking at is the current state of Warp.h / Warp.cpp. Currently, I've added two extra layers of caching: positionPlug is an image containing sample positions, created by sampling the Engine at every point in the data window of the out plug, and positionsAndDerivativesPlug, which is the same thing, but including the 4 partial derivatives in x/y that are used for computing a filter parallelogram.

> What is likely expensive: I've kept that outPlug depends on the hash of the sampler. This means computing the input window of the transform for the tile. That requires computing the value of positionsAndDerivativesPlug, which means computing the value of positionPlug in 5 tiles so that I have adjacency data for the derivatives, just to get a hash. ( It probably doesn't help that I have to hit positionsAndDerivativesPlug 6 times for all the different channels ). I wonder if it would be better to just use a pessimistic hash that says that all output tiles depend on the hash of all input tiles, which would save all this computation happening in the hash?

My suspicion would be that the pessimistic hash of all the input tiles will be far worse for performance. It's unfortunate that we'd have to actually compute the derivatives to make the hash, but once the derivatives are in the cache we should be quicker. So I'd suggest trying some of the other options first.

>Things I'm thinking about trying:

 >a) Decide that in practice we probably won't have analytic warps that need to manually compute derivatives. The numeric derivatives are probably fine. This means that positionsAndDerivativesPlug is purely an internal cache, and doesn't need to make sense to anyone outside ( meaning I should probably put a "__" in the name and make the accessor protected. The first thing that this means is that ensureMinimumParallelogramWidth can be done on the derivatives before they are stored in the cache, instead of considering the cache to be something that users of Gaffer could put potentially unsafe data in, and ensuring this in both hash and compute.

I like the sound of this, mostly from the perspective of keeping the Warp::Engine API around. The Engine layer isolates derived classes from the internal machinery of the Warp and makes it easy to implement new subclasses. It'd be a pity to lose that.

Are we even sure that computing the derivatives is actually expensive enough that they need to be cached? That seems to be a conclusion we jumped to somewhere along the line, and which is currently in danger of dictating everything.

> b) Convert positionsAndDerivativesPlug into not being an image plug. It is something like an image plug, in that it returns a tile sized array of floats that depends on "image:tileOrigin" in the context, but if it's just an internal cache, then it doesn't need to be an image, which means it could return all data at once, rather than requiring 6 separate calls with different context settings for the channel.

Sounds good, assuming we need the cache. I presume the position plug would similarly not be an ImagePlug?

> c) If we don't decide to do a fully pessimistic hash where all input tiles affect all output tiles, then the input window bounds should also be stored on positionAndDerivativesPlug along with the per pixel data, avoiding a call to filteredTileBound that traverses all the pixels inside the hash ( which seems really bad )

Exactly - I suspect that the `filteredTileBound()` call in the hash may be a big part of the problem.

> I'm not sure what the type of positionAndDerivativesPlug should be if you agree it doesn't need to be an image? It could be a CompoundObject, but unnecessary hash map indexing feels a bit wasteful? If it's purely internal, I guess it could just be a FloatVectorData of length "6 * tileSize * tileSize + 4 " ( for 6 channels of data and the bounds ), but I can't imagine you'd like the obscurity of that. It could be a custom struct type, but turning that into Cortex TypedData so I can stuff it in TypedPlug would be a bit verbose. Maybe CompoundObject is the right answer.

I think I'd go for the CompoundObject - it's more readable than the packed FloatVectorData, and with none of the boilerplate of the TypedData approach. Indexing a tiny map a couple of times per tile is going to be negligible compared to all the other work. Maybe the packed approach might improve locality if you interleaved channels, but even then I think that should be packed as one field in the CompoundObject, so the bounds field can be stored separately for reasons of sanity. I'd go with the dumb CompoundData approach and see where we end up - I think the gross performance implications of the hashing strategy are probably worth getting right before we worry about interleaving or not. But...

One other possibility occurs to me :

d) Keep `Engine::inputWindow()`. The great thing about this was that it would allow many engine implementations to return a bound without even needing to compute all the internal positions (think of a noise offset - we know what the magnitude is and can expand by that). Can we simply union the input windows of the surrounding tiles to account for the derivatives they introduce at the edges? It's slightly pessimistic, but nowhere near as pessimistic as hashing the entire input image. And it maintains the possibility of compute-less hashing for a whole class of Warps. If this seems feasible to you, I think it might be worth investigating. I guess maybe there could be a problem because we need to take the filter width into account too?

> The other overall question on Warp is whether the 3 separate cache layers ( enginePlug, positionPlug, positionAndDerivativePlug ) are all necessary. Some thoughts - part of the reasoning you mentioned for Engine is to allow in the future using manipulators through a Warp. This is only going to make sense for very well-behaved warps - maybe for a lens distort, or a rotate if we chose to do rotation this way, but not for some weird arbitrary UV warp that maps one source pixel to 12 different places in the output. To me, in practice, there are two kinds of Warps:

>analytic warps : can be defined by a single engine for all tiles. May support viewing upstream manipulators
pixel defined warps: depend on input image data. Vary per tile. Probably not invertible, and probably not useful to support upstream manipulators. Easiest to define by overriding compute() on positionsPlug
At this point, it feels like maybe the base Warp class is for pixel defined warps, and then maybe a small subclass for analytic warps that adds Engine to automatically drive positionsPlug, and can support your future manipulator trickery. I'm just brainstorming here.

I like the insulation the Engine provides between Warp and its subclasses, and the fact that we can totally reimplement the Warp internals without affecting derived classes at all. It also seems that d) has the potential to mean that the original Engine API continues to be sufficient even when we need derivatives. So it seems a real pity that the caching-the-positions-and-derivatives approach we're currently taking is being used even for analytic warps, where it might even be that repeating a bit of math per pixel is cheaper than hitting memory all the time. And it's a pity to make the extra layers of caching central to the API. So I'm going to suggest that the Engine should stay as the only visible-to-derived-classes part of the API...

danieldresser commented:
I've gotten a bit rabbit-holed on this.

I have gone ahead and put up some commits from Friday which use the new FilterAlgo stuff in Resample, which I could use some feedback on.  There's some renaming needed, with a big comment on the commit about what I think might need to be done.

The next step is performance, which is quite bad, and not for the reasons I was expecting.  I still want to do the cleanup I was thinking of in terms of making positionsAndDerivativesPlug a CompoundObject, and including the bounds there - but after profiling, there doesn't seem to be any substantial time savings there.  The biggest cost right now is all in evaluating the filter kernel - it doesn't help that it looks like we haven't yet picked up this commit from a year ago:
https://github.com/OpenImageIO/oiio/commit/b3962334f3d63272de58d5e6b1411036c993eb9e
... where OIIO improved the performance of some of their filters - in particular the blackman-harris filter that is probably our best candidate for a default.

But even manually hacking up a quick version, I'm still running really slowly - 3 seconds for a 2K frame.  I had convinced myself this was horrifying, but maybe it's just late at night.  Nuke feels a whole lot faster, but this may just be because I wasn't actually forcing a full refresh in Nuke when I was quickly playing with it.

Anyway, it looks slow enough that improving performance would make it more usable, and the main thing that would make it faster seems like it might be computing all the channels at once.  Andrew points out that there is precedent for this in ColorProcessor, but it does feel like it doesn't fit very well with how Gaffer wants me to do things.  Maybe things will make more sense in the morning.  I'm still a bit concerned that I may have just done something totally wrong, and my performance is terrible for no good reason.

danieldresser commented:
Oh, wait, maybe this has a really obvious answer - just fit an axis-aligned bbox to the filter parallelogram instead of filtering properly like it looks like other comp software is doing.  When I first thought about it, I figured that the bbox would hit more pixels than the actual parallelogram, so why not just do it properly, but if the bottleneck is evaluating the filter kernel, that this could go way faster for axis aligned separable filters.  Curious what the visual impact will be, but now I know that I'm trying tomorrow, and wishing I'd tried this sooner.

danieldresser commented:
After having a very tough time not getting pulled in a lot of different directions by the stuff this touches, I'm at a point where I think I would be not unhappy if this ended up looking mostly like this set of commits.

Still not happy about UVWarp needing to use an Engine to convert from an ImagePlug to an Engine back into an ImagePlug, and I still kind of feel like Engines should apply to the whole image, rather than being per tile ( and then we don't the extra layer of caching in positionPlug() to avoid recomputing tiles ), but I can't quite come up with an API that I like, and it didn't sound like you liked the overall idea much, so I guess I'll give up.

Something we agreed on was the positionPlug shouldn't be an image, it should just be a V2fVectorDataPlug which depends on the tileOrigin in the context ( similar to the sampleBoxesPlug ).  Issue with this is that I'm currently passing it into a Sampler - it seems like it would be a shame to replicate the logic in Sampler.  Looking at Sampler, it doesn't actually do anything with the ImagePlug you pass in other than call channelData on it - would it make sense if Sampler just took a FloatVectorDataPlug and the caller would usually call channelDataPlug() on an ImagePlug?  Maybe even template it to work on any vector data plug ( ie V2fVectorDataPlug )?

Aside from that, hopefully this is heading towards being mergeable.  Still some stuff that could probably use renaming ( I just noticed that I don't like sampleBoxesPlug - maybe it should be sampleRegionsPlug ), and some areas where more comments would help.  I kind of want to double check that you don't want any major rearrangements though before I try to finish off the details.

danieldresser commented:
Oh, and yeah - one other thing I need to clean up - hide the internal plugs from the UI once we're happy with how the intermediate plugs work.  The test failures look like it's just the docs failing since the private plugs aren't documented.

Now that I'm trying to actually use this in my original use case, I'm seeing some horribly bad perf.  Haven't investigated yet, but I guess it is very possible that I'm running afoul of the hash cache issues.  Having a tile depend on 5 input tiles in order to get derivatives right on the edges wasn't that bad in simple tests, but I guess if I chain two of those together, and add some upstream processing, I could end up computing the same tiles over and over again?  I guess the test for whether that's my problem is to try running in single-threaded, and if that goes faster, then duplicated upstream calculations are probably the issue.  How are you currently feeling about the improvoments you were working on?

johnhaddon commented on SConstruct:
```diff
@@ -85,7 +85,7 @@ options.Add(
 	"CXXFLAGS",
 	"The extra flags to pass to the C++ compiler during compilation.",
-	[ "-pipe", "-Wall", "-Werror", "-O3" ]
+	[ "-pipe", "-Wall", "-Werror", "-O3", "-ffast-math" ]
```
Do you have any numbers for the improved performance? And since the filters are compiled into OIIO, do you know if we would benefit much from enabling it there?

johnhaddon commented on python/GafferImageTest/UVWarpTest.py:
```diff
@@ -114,5 +115,56 @@ def testNegativeDataWindowOrigin( self ) :
+		uvWarp["filter"].setValue( "blackman-harris" )
+
+		# Enable to write out images for visual comparison
+		if True:
```
`if False` ?

johnhaddon commented on include/GafferImage/FilterAlgo.h:
```diff
@@ -0,0 +1,70 @@
+namespace FilterAlgo
+{
+
+std::vector<std::string> &filters();
```
Should this be returning `const vector &`? And could we call it `filterNames()` since that is what it returns?

danieldresser commented on python/GafferImageUI/ResampleUI.py:
```diff
@@ -81,14 +81,16 @@
 			"description",
 			"""
-			An override for the width of the filter used. This is
+			An override for the scale of the filter used. This is
```
Hmm, yeah, this description is kind of failing.  The default means that the sampled area will effectively be one pixel wide *in output space".  But this doesn't mean that the it will be one pixel wide in input space, or that the filter support will be one pixel wide.  I feel like it's maybe worth mentioning that a cubic with a support of 2, or a box with a support of 1, or a lanczos3 with a support of 6, are all filters that are sampling an area of approximately 1 pixel, just with different sharpness.  But it seems like I've completely failed at communicating this, so maybe I should just remove a bunch of this comment.

johnhaddon replied:
OK, cool - that's good news. I thought maybe I'd got totally the wrong end of the stick about what you were doing. I think I'd be content just to see this described as a multiplier on the filter's natural width, and consider the rest to be internal implementation details. The only client of this plug so far is the Blur node, and that description is just about sufficient to inform the approach taken there.

danieldresser commented:
OK, in theory this should be just about good to go.  Still needs a rebase before merging, but I've left it split out a bit for the moment in case it makes it easier for you to look through.

I think I've hit everything you mentioned, except for your email comment that storing sampleRegion as a Box2f implies to you that it is the entire support width, rather than just the projection of the sample region into input space.  I don't really care much about whether we store sample regions as min/max or as a center point with an offset to the corners, but we don't have a convenient Cortex type for storing a named pair of V2fs do we?

johnhaddon commented on src/GafferImage/Warp.cpp:
```diff
@@ -301,8 +301,8 @@ void Warp::hash( const Gaffer::ValuePlug *output, const Gaffer::Context *context
 
-		filterOverDerivativesPlug()->hash( h );
-		if( filterOverDerivativesPlug()->getValue() )
+		useDerivativesPlug()->hash( h );
```
Since hashing performance is pretty crucial, it's probably worth not calling `hash()` and then `getValue()` for the same plug here. We could just drop the `hash()` call and instead hash in the result of `getValue()`.

danieldresser commented:
OK, I think this ought to do it.  As far as I can see I've addressed all your comments.  The last two need to be squashed in once they're good to go, it's late enough that I want to look over them myself tomorrow morning.

Only contentious thing might be the first commit?  I modified assertImagesEqual to accept images that differ only in the order of channelNames.  I'm assuming that channelNames being stored as a vector is an implementation detail, and conceptually they are a set?  Currently, the cortex object source loads channels in B G R order, which means they can never compare equal to an image loaded through ImageReader.

johnhaddon commented:
> OK, I think this ought to do it. As far as I can see I've addressed all your comments. The last two need to be squashed in once they're good to go, it's late enough that I want to look over them myself tomorrow morning.

OK cool, let me know when you're done with that and I'll merge. I might tweak the FilterAlgo naming slightly after the fact, but won't change anything functional.

> Only contentious thing might be the first commit? I modified assertImagesEqual to accept images that differ only in the order of channelNames. I'm assuming that channelNames being stored as a vector is an implementation detail, and conceptually they are a set? Currently, the cortex object source loads channels in B G R order, which means they can never compare equal to an image loaded through ImageReader.

Hmm, that one's a bit tricky. Currently the order is somewhat important in terms of how things are presented in the UI, and the implementations of several nodes do make an effort to preserve order. So I would prefer that we don't weaken the tests. Since you're testing filtering code, perhaps it's sufficient to do it on a single channel image?

danieldresser commented:
OK, all squashed and hopefully good to go.

For the channel order issue, I did some hackery with a CopyChannels in the test case to reorder the channels coming out of ObjectToImage.  Feels hacky, but it's just for defining a source image for the test case.  Would be nice if I could use OSLImage for this.  Maybe I should have just added an extra source image as an exr.

johnhaddon commented:
Cool - thanks!
